rules/handler.go

```diff
@@ -0,0 +1,391 @@
+
+	// Ruleset Errors
+	errNoUpdate      = errors.New("no new config given")
+	errNotTombstoned = errors.New("not tombstoned")
+
+	ruleActionErrorFmt    = "cannot %s rule %s. %v"
+	ruleSetActionErrorFmt = "cannot %s ruleset %s. %v"
+)
+
+// HandlerOptions holds common settings to all handlers.
+type HandlerOptions struct {
+	propagationDelay time.Duration
+}
+
+// NewHandlerOpts creates a new HanlderOptions struct.
+func NewHandlerOpts(propDelay time.Duration) HandlerOptions {
+	return HandlerOptions{propagationDelay: propDelay}
+}
+
+// MappingRuleConfig is the parts of a mapping rule that an end user would care about.
+// This struct is provided to create or make updates to the mapping rule.
+type MappingRuleConfig struct {
+	Name     string            `json:"name" validate:"nonzero"`
+	Filters  map[string]string `json:"filters" validate:"nonzero"`
+	Policies []policy.Policy   `json:"policies" validate:"nonzero"`
+}
+
+// MappingRuleUpdate contains a MappingRuleConfig along with an ID for a mapping rule.
+// The rule with that ID is meant to updated to the given config.
+type MappingRuleUpdate struct {
+	Config MappingRuleConfig `json:"config" validate:"nonzero"`
+	Name   string            `json:"name" validate:"nonzero"`
+}
+
+// RollupRuleConfig is the parts of a rollup rule that an end user would care about.
+// This struct is provided to create or make updates to the rollup rule.
+type RollupRuleConfig struct {
+	Name    string            `json:"name" validate:"nonzero"`
+	Filters map[string]string `json:"filters" validate:"nonzero"`
+	Targets []RollupTarget    `json:"targets" validate:"nonzero"`
+}
+
+// RollupRuleUpdate is a RollupRuleConfig along with an ID for a rollup rule.
+// The rule with that ID is meant to updated with a given config.
+type RollupRuleUpdate struct {
+	Config RollupRuleConfig `json:"config" validate:"nonzero"`
+	Name   string           `json:"name" validate:"nonzero"`
+}
+
+// RuleSetHandler exposes all of the methods to work with a RuleSet
+type RuleSetHandler interface {
+	// AppendMappingRule creates a new mapping rule and adds it to this ruleset.
+	AddMappingRule(MappingRuleConfig) error
+
+	// UpdateMappingRule creates a new mapping rule and adds it to this ruleset.
+	UpdateMappingRule(MappingRuleUpdate) error
+
+	// DeleteMappingRule deletes a mapping rule
+	DeleteMappingRule(string) error
+
+	// AppendRollupRule creates a new mapping rule and adds it to this ruleset.
+	AddRollupRule(RollupRuleConfig) error
+
+	// UpdateRollupRule creates a new mapping rule and adds it to this ruleset.
+	UpdateRollupRule(RollupRuleUpdate) error
+
+	// DeleteRollupRule deletes a rollup rule
+	DeleteRollupRule(string) error
+
+	// Tombstone tombstones this ruleset and all of its rules.
+	Delete() error
+
+	// Revive removes the tombstone from this ruleset. It does not revive any rules.
+	Revive() error
+
+	// Validate checks that this ruleset is good to use. E.i. not tombstoned
```

Comment:
lol is `E.i.` some newspeak that I'm unfamiliar with or should it be i.e.?

You may not want to indicate what all you're checking here, it'll keep the comment accurate as your implementation could check additional things (e.g. if an add/update operation left the underlying ruleset in a bad state).